### ISSUE TITLE
No Host Name option (12) in DHCP request

### ISSUE THREAD

**Issue by agners** [esphome/issues#2057]
**Operating environment/Installation (Hass.io/Docker/pip/etc.):**

Running ESPHome in Dev-Container

**ESP (ESP32/ESP8266, Board/Sonoff):**

ESP32 LILYGO TTGO T1

**ESPHome version (latest production, beta, dev branch)**

v1.17.2 and 1.19.0-dev.

**Affected component:**

wifi

**Description of problem:**

**Problem-relevant YAML-configuration entries:**
```yaml
esphome:
  name: neopixel16
  platform: ESP32
  board: ttgo-t1

wifi:
  ssid: [redacted]
  password: [redacted]

logger:

api:
  password: [redacted]

ota:
  password: [redacted]
```

(I usually also have `light` and `status_led` in that YAML, but it seems to behave the same way without the two)

**Logs (if applicable):**

```
[12:17:18][I][app:105]: ESPHome version 1.19.0-dev compiled on May 16 2021, 12:16:49
[12:17:18][C][wifi:443]: WiFi:
[12:17:18][C][wifi:303]:   SSID: [redacted]
[12:17:18][C][wifi:304]:   IP Address: 192.168.80.173
[12:17:18][C][wifi:306]:   BSSID: [redacted]
[12:17:18][C][wifi:307]:   Hostname: 'neopixel16'
[12:17:18][C][wifi:311]:   Signal strength: -63 dB ▂▄▆█
[12:17:18][C][wifi:315]:   Channel: 11
[12:17:18][C][wifi:316]:   Subnet: 255.255.255.0
[12:17:18][C][wifi:317]:   Gateway: 192.168.1.1
[12:17:18][C][wifi:318]:   DNS1: 192.168.1.1
[12:17:18][C][wifi:319]:   DNS2: 0.0.0.0
[12:17:18][C][logger:189]: Logger:
[12:17:18][C][logger:190]:   Level: DEBUG
[12:17:18][C][logger:191]:   Log Baud Rate: 115200
[12:17:18][C][logger:192]:   Hardware UART: UART0
[12:17:18][C][ota:029]: Over-The-Air Updates:
[12:17:18][C][ota:030]:   Address: neopixel16.local:3232
[12:17:18][C][ota:032]:   Using Password.
[12:17:18][C][api:095]: API Server:
[12:17:18][C][api:096]:   Address: neopixel16.local:6053
```

Wireshark capture
```
Frame 2547037: 350 bytes on wire (2800 bits), 350 bytes captured (2800 bits) on interface /tmp/wireshark_extcap_sshdumpC5VK30, id 0
Ethernet II, Src: Espressi_24:cc:14 (a4:cf:12:24:cc:14), Dst: Broadcast (ff:ff:ff:ff:ff:ff)
Internet Protocol Version 4, Src: 0.0.0.0, Dst: 255.255.255.255
User Datagram Protocol, Src Port: 68, Dst Port: 67
Dynamic Host Configuration Protocol (Request)
    Message type: Boot Request (1)
    Hardware type: Ethernet (0x01)
    Hardware address length: 6
    Hops: 0
    Transaction ID: 0x3a04399c
    Seconds elapsed: 0
    Bootp flags: 0x0000 (Unicast)
    Client IP address: 0.0.0.0
    Your (client) IP address: 0.0.0.0
    Next server IP address: 0.0.0.0
    Relay agent IP address: 0.0.0.0
    Client MAC address: Espressi_24:cc:14 (a4:cf:12:24:cc:14)
    Client hardware address padding: 00000000000000000000
    Server host name not given
    Boot file name not given
    Magic cookie: DHCP
    Option: (53) DHCP Message Type (Request)
    Option: (57) Maximum DHCP Message Size
    Option: (50) Requested IP Address (192.168.1.173)
    Option: (55) Parameter Request List
    Option: (255) End
    Padding: 000000000000000000000000000000000000000000000000000000000000000000000000…
```

**Additional information and things you've tried:**
My router (OpenWrt) by default adds host names/IPs to a local zone (`.lan`). Without hostname it can't do that. Hence I cannot use DNS manged by the router.

I am aware that mDNS is an option and this works, however mDNS comes with its own set of problems (resolving is sometimes delayed, lack of reliability). So having the option to resolve ESPHome hosts using regular DNS would be nice.

**Comment by agners** [comment]
Issue #1855 seems related.

**Comment by agners** [comment]
It seems that this is rooting in a LWIP issue used in ESP IDF, probably best described in this pull request: https://github.com/espressif/esp-lwip/pull/6

From what I can see this has been resolved with the following commit sometime between ESP IDF release 3.3.4 and 3.3.5:
https://github.com/espressif/esp-idf/commit/668347e9c153819233b4ec944530b5b4e943ad0a

And indeed, using Arduino Core to 1.0.5 (platformio/espressif32 @ 3.1.1, which seems to use a non-released version sometime between ESP IDF 3.3.4 and 3.3.5 and apparantely containing the fix) or 1.0.6 (platformio/espressif32 @ 3.2.0, which uses ESP IDF 3.3.5) seems to resolve the issue:

```yaml
esphome:
  name: neopixel16
  platform: ESP32
  board: ttgo-t1
  arduino_version: 1.0.5
```
